### ISSUE TITLE
Clear speech viewer

### ISSUE THREAD

**Issue by jmuheim** [nvaccess/nvda#6582]
I'm recording tutorials for NVDA as video. It would be nice to clear the speech viewer before starting recording. Is there a way to do this?

**Comment by feerrenrut** [comment]
No, not directly. As a workaround you could fill the visible area with white space. Do this by opening notepad, pressing enter, then with the cursor on the first line keep pressing the left arrow. This should report as "carriage return" and create a blank line.

**Comment by feerrenrut** [comment]
Since clearing the speech viewer is quite an edge case, that is unlikely to apply to most users I'm going to close this issue.

**Comment by jmuheim** [comment]
I agree this is an edge case.

I think I need this because I want to create movies of how to use NVDA with speech viewer open. But because it's always opened somewhere in the middle of the screen, I have to move it with the mouse and then go back to my browser. This fills the speech viewer with a lot of irrelevant output already, and I want to get rid of that so the speech viewer is nice and clean when starting the record.

I had proposed to [save the speech viewer's position when closing it and moving it to the old position when reopening it](https://github.com/nvaccess/nvda/issues/5050) before already. Sadly, though [this issue]https://github.com/nvaccess/nvda/pull/6202) claims that this feature was implemented, it doesn't work for me (NVDA 2016.3), am I missing something here?

So I hoped I could do stuff manually like this.

feerrenrut's hint is useful, but still, when switching back from notepad to the browser, there is already output produced again.

**Comment by feerrenrut** [comment]
Saving the speech viewer position will be included with 2016.4. If you wish to use this feature now you could try out a [nightly build](http://www.nvaccess.org/files/nvda/snapshots/) (scroll to the "master" section). 

For the following I will assume you are using the desktop keyboard layout when specifying keyboard shortcuts. If not, please check the user guide for equivalent keys for the laptop keyboard layout.

Yes, switching back to chrome will fill the speech viewer with text. Using "sleep mode" will reduce the amount of text, however some is still shown. To do this:
- With Chrome in focus
- Enable "sleep mode" (`nvda+shift+s`) in Chrome. Sleep mode is application specific.
- Use `alt+tab` to switch to the notepad window
- Clearing the speech viewer using the method outlined previously (by repeating the "carriage return" character)
- Use `alt+tab` to switch back to Chrome.

If you are not demoing the review cursor, the following may work for you:
- In NVDA options, mouse settings, ensure "mouse tracking" is disabled.
- `alt+tab` to notepad
- position the cursor so that `numpad down arrow` (`numpad 2` with `num lock` off) will report "carriage return" and add a blank line to the speechviewer
- Press `nvda+6` then `nvda+7` to turn off "caret moves review cursor" and "focus moves navigator object" respectively.
- `alt+tab` back to chrome
- press `numpad down arrow`, "carriage return" should continue to be reported and you should be able to clear the speechviewer.

Please let us know if these steps allow you to do what you need. If not this feature would be best if implemented as a global plugin, rather than a core feature. Since a keyboard shortcut to do this would rarely be applicable.

While we appreciate that this would be useful to you and we wish to try to support the creation of NVDA tutorials, we need to prioritise this against other issues. I will set priority 4 for this issue, which unfortunately means its unlikely to get done by the NVAccess developers. You are welcome to ask in the NVDA developer mailing list, perhaps a community member would be willing to implement this for you. If you have a tight deadline for this and need this feature developed quickly please contact us through admin@nvaccess.org to discuss paid services we are able to provide.

**Comment by jmuheim** [comment]
> Saving the speech viewer position will be included with 2016.4. If you wish to use this feature now you could try out a nightly build (scroll to the "master" section).

This is good news. Thank you!

**Comment by jmuheim** [comment]
> Saving the speech viewer position will be included with 2016.4.

Just downloaded 2016.4, but couldn't find a way to make Speech Viewer re-opening at the previous position. Is there an option I have to activate?

**Comment by feerrenrut** [comment]
No, there is no option that must be enabled to get this behaviour. The speech viewer should attempt to open with the same dimensions and at the same location. The option to automatically open when NVDA starts is at the bottom of the speech viewer window.

The speech viewer is not able to open in the same location if the number of screens, or the screen resolution has changed since the last time it was opened. 

If this this still does not work for you, could you please open a new issue on github and attach a log from NVDA?

**Comment by jmuheim** [comment]
It seems that the new version wasn't installed yet, although I clicked "Download and install update" this morning (maybe something interrupted installation). I did it manually again, and it works now. Sweet! :-)

**Comment by iamrauu** [comment]
Here's a workaround:
Configure a keyboard shortcut that toggles the Speech viewer on/off. 

**Steps**
- From context menu, show Preferences then choose Input Gestures.
- Expand Tools, select "Toggles the NVDA Speech viewer..." and choose Add.
- Press the keyboard combination that you'd like to assign for this option.
- Choose apply to _keyboard desktop or keyboard, all layouts_; then.
- Save and voila!

Hope this helps :smile:

**Comment by bhavyashah** [comment]
@jmuheim  @feerrenrut @iamrauu What is the final conclusion with regards to the UX of this ticket? Do we want an option to clear the contents of the speech viewer, do we want an unbound assignment for toggling the speech viewer, or something else? A summary would be helpful to understand the scope and impact of this feature suggestion.

**Comment by feerrenrut** [comment]
I think [the suggestion](https://github.com/nvaccess/nvda/issues/6582#issuecomment-276759628) by @iamrauu is adequate.  This is quite a specialised request, and does not really need a to be a built in keyboard shortcut. As a note to anyone thinking about adding a button on to the speech viewer to do this, one problem you will face is that there will be a focus switch when pressing the button.

**Comment by jmuheim** [comment]
The suggestion by @iamrauu works for me. I don't need a specific feature for this. I close the issue.

**Comment by SammyJoeOsborne** [comment]
As a developer, I would love to have the option to clear the speech viewer. It is especially useful for testing. It would be nice to be able to clear the viewer before I performed a particular action on a webpage, to see the output that particular action generated.

**Comment by feerrenrut** [comment]
@SammyJoeOsborne Did you try the [the suggestion](https://github.com/nvaccess/nvda/issues/6582#issuecomment-276759628) from iamrauu?

**Comment by shard2001** [comment]
Just another developer here adding that a simple clear button on the window would be a help for testing. The Notepad way does work, but it would be easier
